### ISSUE TITLE
Changes to share options

### ISSUE THREAD

**Issue by benetechMatt** [benetech/MathShare#1086]
In an effort to make it easier for students to turn in answered problem sets and for teachers to share sets with students AND with other teachers we are making these changes to the sharing process.

Currently there is a row of buttons for sharing work (permalink, google classroom, microsoft teams).  These have different URLs and actions based on whether the user is on the /edit /solve /answered versions of the problem page.

Changing this to two buttons that all users have and link out to different URLS

"Share My Answers" is meant for students that are sharing their answered problem sets.  These share the view only /review/ link.
When users select the "Share my answers"

![Mockup](https://user-images.githubusercontent.com/49413678/77803100-5a1d9380-704a-11ea-9a37-6e187c550c47.png)

**Comment by jscholes** [comment]
We haven't been assigned for a11y review as yet, but "what's this?" is not a descriptive link label.  I want to surface that now so that an alternative can be decided on before we review it properly.

**Comment by alexrcabral** [comment]
@benetechMatt Isn't this changed from the call this morning? 

Separate section for problem set creators with problem set level controls. A three dot menu on the problem set (for all users) which contains the share problem set option.

@jscholes & @benetechMatt Would a link labeled "Sharing FAQ" be more appropriate?

**Comment by benetechMatt** [comment]
Rework of this issue. There will be two views of the problem set page.  Creators, or owners of the problem set page have new "problem set controls" panel which allows them to make changes to the set and share an answerable version of the set.

![image](https://user-images.githubusercontent.com/49413678/78508061-b6b73780-7749-11ea-8e7d-2714a9eee53f.png)


Users that did not create the set or are not the owners of the set do not see the problem set controls section.  They are able to answer questions, share their answers, and duplicate the set.

![image](https://user-images.githubusercontent.com/49413678/78508069-c171cc80-7749-11ea-9762-c86ed33a8c57.png)


![image](https://user-images.githubusercontent.com/49413678/78508076-cc2c6180-7749-11ea-8ea6-2071d2586599.png)

**Comment by alexrcabral** [comment]
Clarifying question: 

> "Require description for each step" and a check box next to the label. When checked, anyone that completes the set must add an explanation to each step.

Is this based on what was checked at the time of assignment, at the time the student problem set is accessed (when a copy is made), or is it a global control which will retroactively make any students' sets tied to the teacher's set require (or not require) an explanation?

**Comment by benetechMatt** [comment]
Splitting up this issue into three
1. This issue is just about the new method for sharing problem sets
2. Ability to make step descriptions optional #1094 
3. Add problem set controls #1129 

The buttons, language, and links used for sharing are different for problem set creators/owners and for those that are viewing/answering the set.

### Creators/owners of the set


### Non-owners of the set
Modal with the heading "Share My Answers" and a checkmark icon on the left
bulleted list explaining what it's for

- Allows others to see your work
- Your work cannot be edited
- The problem set cannot be duplicated

"Submit your answers to Google Classroom or Microsoft Teams or you can also share your finished work with the link below."

The three sharing options are listed in rows below this.

URL of the /review/ link and a button with the label "copy" to the right of it.
Google Classroom button.  works the same as the current version
Microsoft Teams button. works the same as the current version

![mockup of the "share by answers" modal.](https://user-images.githubusercontent.com/49413678/79018894-91a43980-7b3a-11ea-915f-7e75f5a09053.png)

**Comment by alexrcabral** [comment]
@benetechMatt One of the identified necessary functions was for individuals to share sets they did not create. Share problem set should not be located within the problem set controls because only problem set creators can access this. This should instead be within the context menu of the problem set, which can be accessed by anyone with the problem set.

**Comment by benetechMatt** [comment]
@alexrcabral users must duplicate the set before they can share the problem set. This is why it's located in the controls box.  Once they duplicate the set they will be the owner and then see the "share problem set" button.
Users should not be able to share the problem set unless they are the owner of it.

**Comment by alexrcabral** [comment]
@benetechMatt Why is it important for a user to not be able to share a problem set if they are not the creator of the problem set?

**Comment by alexrcabral** [comment]
Decision on call: you cannot share a set that you did not create or duplicate. 
This is intended to reduce possibility of error for students sending the wrong link, and to restrict control in assigning sets.

**Comment by benetechMatt** [comment]
Adding issue #1131

**Comment by benetechMatt** [comment]
Here is a summary of the final solution decided on. Details for the updated "share problem set" and "Share my answers" buttons in comment https://github.com/benetech/MathShare/issues/1086#issuecomment-612191139

This issue is just about the new method for sharing problem sets
- Ability to make step descriptions optional #1094
- Add problem set controls #1129
- change wording to duplicate problem set menu option #1130
- Creators can create a key for problems / problem set #1131

Problem set page for set owners. This shows the changes detailed in the above issues for users that are unauthenticated and the "owner" of the set.
![Mockup of the problem set page for owners. "Share My Answers" button to the right of the set title. set controls box has been added below the title. It has a blue outline and light gray background.](https://user-images.githubusercontent.com/49413678/79755643-10d60200-82df-11ea-8eef-5ab96c28fcb7.png)

Problem set page for users that are solving the set. This shows the changes detailed in the above issues for users that are unauthenticated and either solving the set (student) or viewing it before duplicating it (other educators).
![mockup for students. No problem set controls](https://user-images.githubusercontent.com/49413678/79756181-db7de400-82df-11ea-8507-908e3bfc5da1.png)

**Comment by rupeshparab** [comment]
The mockups in this change has some overlap with https://github.com/benetech/MathShare/issues/1129

For now I have not added the UI of problem set controls in this change, so the share button for teachers is on the right corner.

Also, the warning change is not present in the current branch, it is being tracked here: https://github.com/benetech/MathShare/issues/1073

**Comment by johnhbenetech** [comment]
@alexrcabral In the final set of mockups the 'Share My Answers' button appears in the teacher view. This doesn't really make sense technically so we assume it was included in error. Let us know if we are misunderstanding something

**Comment by jscholes** [comment]
I'm unclear about the workflow of this ticket.  We haven't been assigned to carry out accessibility review on this, nor on any of the related ones.  But it has been moved to development, and implemented with accessibility problems which we should've had a chance to address before code was written.

I have explicitly reverted PR https://github.com/benetech/MathShare/pull/1181.  Please provide more insight into why this was developed without us signing off on it, and allow us to do so before a new PR is prepared.

CC @johnhbenetech, @benetechMatt, @alexrcabral.

**Comment by johnhbenetech** [comment]
@alexrcabral @benetechMatt This was moved to the Ready for Development column so I assumed it had passed review. I won't have the bandwidth to double check that everything has gone through the prescribed process so I rely on that workflow being accurate.

@alexrcabral can you pull this back and make sure to discuss it with the a11y team before we receive it for reimplementation

CC @jscholes

**Comment by jscholes** [comment]
There are several problems to pick apart here (in addition to the workflow concerns):

1. "Require description for each step" cannot be a check box, if it causes a setting or some other state to be updated without an apply-style button being pressed first.  Instead, it must be a `<button>` with an `aria-pressed` attribute.
2. For the "Share Problem Set" modal, @benetechMatt has suggested a level 3 heading.  There are several reasons why this is problematic:
	1. All modals must begin with an H1.
	2. The design phase shouldn't be concerned with heading levels, full-stop.  Hierarchy and structure, absolutely.  But the levels are not a design decision.
	3. Further to point `#2`, headings are for structure, not visuals.
3. On certain views, the only options available to a user are to duplicate or share the set.  However:
	1. These options aren't positioned together.  The duplication control is behind the options dropdown toggle, but the button to invoke the sharing modal is underneath in the problem set controls/actions section.
	2. Because there are no other actions available other than to share the set, screen reader users encounter a list of one item which is semantically useless.  Same goes for the options dropdown; having two controls that only have one item seems redundant.
4. The control to copy the permalink doesn't reuse any of the mark-up used for https://github.com/benetech/MathShare/pull/1177, so the experience is inconsistent and shares the accessibility problems outlined on that PR.  This should be extracted out into a reusable component.  As an aside, we are still awaiting comments on that PR re: analytics before it can be merged.
5. Why do sharing-related controls use a check mark?  They don't toggle anything.

CC @alexrcabral and @benetechMatt

**Comment by sinabahram** [comment]
Please note, there are other issues as well that this lack of process is true about. James and I expressed extreme and unmitigated concern about this on our call with Alex yesterday on this topic.

 

If the process is not followed, then there “will”, not “may”, be accessibility and other issues. This is a guarantee, not a possibility, and it is a P0 risk element in my opinion.

 

In short, without exception, this should never be allowed to happen again, full stop.

**Comment by alexrcabral** [comment]
@johnhbenetech 
That is not an error. It's meant to support students who are creating their own problem sets being able to share their answers without having to duplicate their own set.

@jscholes 
1. #1094 addresses this. The control is surfaced as toggle button with aria-pressed.
2. We won’t discuss headings since that’s not a design decision. However, if you could give suggestions or examples for heading structure for Rupesh on the development side that would be helpful. 
3. While both the problem set level controls and the options are very sparse right now, they will be filled out with time. For ease of use, we will move the “Share Problem Set” button to the options dropdown underneath the option to duplicate. This will only be listed as an option for problem set owners. 
4. This I’ll leave to @rupeshparab since it concerns specifically the development portion.
5. I’m a bit unclear on your question so apologies if this isn’t what you were asking. The controls use a toggle rather than a checkmark, and they determine whether the accompanying work is shared along with the prompts, as well as whether work will be required on any sets generated from the share function.

@sinabahram Agreed. All scoped issues will be going through the Accessibility Review touchpoint for design review moving forward.

**Comment by jscholes** [comment]
> if you could give suggestions or examples for heading structure for Rupesh on the development side that would be helpful.

Start the modal with an H1, H2 for child headings, H3 for grandchild headings, etc.  I only expect there to be an H1 in this modal, but this is the same heading-related advice that applies everywhere.

> While both the problem set level controls and the options are very sparse right now, they will be filled out with time.

Understood, but it depends what that timeline looks like.  If we're talking a couple of weeks before a second option is added, I'm fine with it being a single-item list until then.  If we're talking six months, I'm not.  But overall, the accessibility mindset should be as follows:

* When we have a single item, mark it up appropriately, or add it to an existing control like a dropdown.
* As soon as we have more than one, it is a relatively small amount of effort to add list mark-up or undo a previous move.

> The controls use a toggle rather than a checkmark

I'm unclear what we're talking about here.  I'm referencing this part of the mock-up/design description:

> there is a button with a check mark icon on the left and label of "Share My Answers".

It's phrased as an action and prefixed with a verb, so what is the checkmark for?  Apologies if I'm misunderstanding something here.

**Comment by alexrcabral** [comment]
@jscholes 
> It's phrased as an action and prefixed with a verb, so what is the checkmark for?

Thanks for clarifying, I get the issue now! This isn't a checkbox control nor is it visually similar to a checkbox control. This is a visual icon as part of the button for ease of understanding. It's modeled after a simple checkmark similar to the "Finished" icon, for students to understand it's the next step when they're done with their set. 

> depends what that timeline looks like

Absolutely. Since in this case it's not currently slated for the next few weeks and will likely be a couple months out, the move of the "Share Problem Set" button to the dropdown control will help in the interim so there is no single-item list. 

Also a mockup of the problem set owner's view for clarification @rupeshparab: 

![Mockup of the problem dropdown. The three button control is highlighted showing the dropdown containing the options "Duplicate Set" and "Share Problem Set"](https://user-images.githubusercontent.com/47576069/82256876-f27a1b00-990b-11ea-96ef-0f785d07e417.png)

This will only show up for problem set owners ("edit" link) and will no longer be displayed in the "Problem Set Controls" box. The problem set solver view ("solve" link) will show the first option with the new copy of "Duplicate to edit/share" but will **not** contain Share Problem Set within the dropdown.

**Comment by johnhbenetech** [comment]
@alexrcabral I'm thoroughly confused since there's so much going on across various issues. Can you give us updated mockups for the problem set controls/three dots for each of the user types and cases. There are a lot of combinations here and I fear it's ripe for misunderstanding on our part

**Comment by jscholes** [comment]
@alexrcabral All good, thanks for that input.

**Comment by alexrcabral** [comment]
@johnhbenetech I can get those in tomorrow morning / early afternoon since the sharing modal needs to be remade for the edits.

**Comment by alexrcabral** [comment]
@johnhbenetech @rupeshparab This may be repetitive but hopefully this will clarify what's being asked in this issue. If you have any questions please don't hesitate to ask.

### Creators/owners of the set only
Problem set creators/owners can share a version of the set that can be completed by others. This is done through a "Share Problem Set" button located in the problem set dropdown. 

<img width="676" alt="Mockup of edit view, with three dot dropdown selected. Beneath are two options, Duplicate and Share Problem Set." src="https://user-images.githubusercontent.com/47576069/82376085-74367b00-99d6-11ea-8de6-2ac1bba69a49.png">

The dropdown contains the options with their accompanying icons: 
1. "Duplicate"
2. "Share Problem Set"

On choosing to Share Problem Set, the modal appears. 

<img width="424" alt="Mockup of Share Problem Set modal, which appears over the greyed out site in the background. It contains a bulleted list, a one sentence description, and the share buttons at the bottom." src="https://user-images.githubusercontent.com/47576069/82376245-a3e58300-99d6-11ea-821e-0f3ba50e2edb.png">

The modal is titled "Share Problem Set" modeled after the button, including the share icon to the left. This is followed by a bulleted list: 
- "Allows students to answer questions"
- "Allows educators to duplicate and re-use this problem set"

Followed by the text, "Share your problem set through Google Classroom or Microsoft Teams, or share using the link below." 

Beneath are two columns of buttons. On the left is the option to "Copy Link URL" with the link icon. On the right column is the "Google Classroom" option and the "Microsoft Teams" option with their accompanying icons. 

If "Copy Link URL" is clicked, it should bring up a notification saying "Successfully copied to clipboard". 


### All users 

All users, including problem creators, have the option to "Share My Answers" to share a view-only version of the set. This is located in the top right of both views, in line horizontally with the problem set title and edit/dropdown controls.

Note: The option to Share Problem Set is *not* included in the dropdown for users in the "solve" view. 

<img width="675" alt="Mockup of solve view, with three dot dropdown selected. Beneath is the option Duplicate to edit/share. On the right of the screen is the button to Share My Answers" src="https://user-images.githubusercontent.com/47576069/82376107-7dbfe300-99d6-11ea-8912-f20c597a74af.png">

When Share My Answers is selected, this brings up the modal titled "Share My Answers." 

<img width="418" alt="Mockup of Share My Answers modal, which is laid out similarly to the Share Problem Set Modal. It also appears over the greyed out site in the background. It contains a bulleted list, a one sentence description, and the share buttons at the bottom." src="https://user-images.githubusercontent.com/47576069/82377110-08eda880-99d8-11ea-8a17-5235cc770a5d.png">

Copy: 
- "Allows others to see your work"
- "Your work cannot be edited"
- "The problem set cannot be duplicated"

"Submit your answers through Google Classroom or Microsoft Teams, or share using the link below."

Similar to the Share Problem Set, this is followed by the two columns of buttons, with the same labels, icons, and interactions (Copy Link URL, Google Classroom, Microsoft teams).

**Comment by jscholes** [comment]
@alexrcabral From our side, summaries like this are incredibly helpful in gaining an understanding of the current state of play for a design or idea.  So no worries about repetition over here.

**Comment by abaranowska1** [comment]
Thanks, @alexrcabral  for sum up the whole discussion in the end, it is helpful during testing :) 

From QA side:

**Tested with success**

**Verified cases:** 
- Problem set creators/owners can share a version of the set that can be completed by others. This is done through a "Share Problem Set" button located in the problem set dropdown.
![Peek 2020-06-10 12-51](https://user-images.githubusercontent.com/61964176/84259744-38945a00-ab19-11ea-91d7-be2072875d41.gif)
- The dropdown contains the options with their accompanying icons:
1. "Duplicate"
2. "Share Problem Set"
- On choosing to Share Problem Set, the modal appears
![image](https://user-images.githubusercontent.com/61964176/84259891-6e394300-ab19-11ea-9ec0-9d9c91be84da.png)
- After click the "Copy Link URL" URL is copied to the clipboard, **but there is no notification saying "Successfully copied to clipboard".** @rupeshparab 
- After click Google classroom or teams button, all works fine 
- All users, including problem creators, have the option to "Share My Answers" to share a view-only version of the set. 
![image](https://user-images.githubusercontent.com/61964176/84260522-7940a300-ab1a-11ea-9027-8a816414b597.png)
- This is located in the top right of both views, in line horizontally with the problem set title and edit/dropdown controls.
- The option to Share Problem Set is not included in the dropdown for users in the "solve" view.
![image](https://user-images.githubusercontent.com/61964176/84260267-0cc5a400-ab1a-11ea-8fd7-7f3e0ec1e68c.png)
- When Share My Answers is selected, this brings up the modal titled "Share My Answers."
![image](https://user-images.githubusercontent.com/61964176/84260390-40083300-ab1a-11ea-812a-5a843883c3b1.png)
- After click the "Copy Link URL" URL is copied to the clipboard, **but there is no notification saying "Successfully copied to clipboard".** @rupeshparab 

**Verified on:**
- Share options verified on the teacher, student, other and unauthenticated user
- QA instance
- Ubuntu 16.04 OS
- Mozilla Firefox v. 77.0.1

**To sum up:** 
@rupeshparab Please add notification saying "Successfully copied to clipboard" after click the "Copy Link URL"  in  "Share My Answers" and "Share Problem Set"

**Comment by rupeshparab** [comment]
@abaranowska1 
The "Successfully copied to clipboard" announcement is currently using aria-live, it should be audible when using a screen reader. Will also add a visual alert for it.

**Comment by abaranowska1** [comment]
@rupeshparab Oh ok, I thought that it should look like a standard notification for example 
![image](https://user-images.githubusercontent.com/61964176/84262764-79db3880-ab1e-11ea-8bc2-afab6a441cd3.png)

**Comment by abaranowska1** [comment]
@rupeshparab 
I didn't see any visual notification after copying link.

![Peek 2020-06-15 09-07](https://user-images.githubusercontent.com/61964176/84628114-5e8a7780-aee8-11ea-8a86-32ca1f317e5f.gif)

- QA instance
- Ubuntu 16.04 OS
- Mozilla Firefox v. 77.0.1
- Google Chrome v. 83.0.4103.61

**Comment by rupeshparab** [comment]
The PR of this change (https://github.com/benetech/MathShare/pull/1244) has not been merged, it is currently in Accessibility Audit
